File: apps/web/open2log/lib/open2log/stores/shop.ex
defmodule Open2log.Stores.Shop do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "shops" do
    # Overture Maps GERS ID for unique identification
    field :gers_id, :string

    # Basic info
    field :name, :string
    field :chain, Ecto.Enum, values: [:lidl, :s_kaupat, :k_market, :tokmanni, :prisma, :other]
    field :address, :string
    field :city, :string
    field :postal_code, :string
    field :country, :string, default: "FI"

    # Location - H3 index for efficient geo queries
    field :latitude, :float
    field :longitude, :float
    field :h3_index, :string

    # Opening hours (stored as JSONB-like map)
    field :opening_hours, :map

    timestamps(type: :utc_datetime)
  end

  def changeset(shop, attrs) do
    shop
    |> cast(attrs, [:gers_id, :name, :chain, :address, :city, :postal_code,
                    :country, :latitude, :longitude, :h3_index, :opening_hours])
    |> validate_required([:name, :chain, :latitude, :longitude])
    |> unique_constraint(:gers_id)
    |> compute_h3_index()
  end

  defp compute_h3_index(changeset) do
    # H3 index will be computed when we have lat/lng
    # Resolution 9 gives ~174m hexagons, good for store identification
    lat = get_change(changeset, :latitude)
    lng = get_change(changeset, :longitude)

    if lat && lng do
      # This would use an H3 library - placeholder for now
      # h3_index = H3.geo_to_h3(lat, lng, 9)
      changeset
    else
      changeset
    end
  end
end

File: apps/ios/Open2Log/Sources/Models/Shop.swift
import Foundation
import CoreLocation

struct Shop: Codable, Identifiable {
    let id: UUID
    let gersId: String? // Overture Maps GERS ID
    let name: String
    let chain: ShopChain
    let address: String
    let city: String
    let postalCode: String
    let country: String
    let latitude: Double
    let longitude: Double
    let h3Index: String?
    let openingHours: [String: String]?

    enum CodingKeys: String, CodingKey {
        case id
        case gersId = "gers_id"
        case name, chain, address, city
        case postalCode = "postal_code"
        case country, latitude, longitude
        case h3Index = "h3_index"
        case openingHours = "opening_hours"
    }

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    var location: CLLocation {
        CLLocation(latitude: latitude, longitude: longitude)
    }

    func distance(from location: CLLocation) -> CLLocationDistance {
        self.location.distance(from: location)
    }
}

enum ShopChain: String, Codable, CaseIterable {
    case lidl
    case sKaupat = "s_kaupat"
    case kMarket = "k_market"
    case tokmanni
    case prisma
    case other

    var displayName: String {
        switch self {
        case .lidl: return "Lidl"
        case .sKaupat: return "S-kaupat"
        case .kMarket: return "K-market"
        case .tokmanni: return "Tokmanni"
        case .prisma: return "Prisma"
        case .other: return "Other"
        }
    }
}
